lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -0,0 +1,760 @@
+                           (self.subjectAltName_strict and not set(l_subjectAltName) == set(str(extension).split(', '))):
+                            self.message.append(
+                                'Invalid subjectAltName component (got %s, expected all of %s to be present)' % (str(extension).split(', '), l_subjectAltName)
+                            )
+
+        def _validate_notBefore():
+            if self.notBefore:
+                if self.cert.get_notBefore() != self.notBefore:
+                    self.message.append(
+                        'Invalid notBefore component (got %s, expected %s to be present)' % (self.cert.get_notBefore(), self.notBefore)
+                    )
+
+        def _validate_notAfter():
+            if self.notAfter:
+                if self.cert.get_notAfter() != self.notAfter:
+                    self.message.append(
+                        'Invalid notAfter component (got %s, expected %s to be present)' % (self.cert.get_notAfter(), self.notAfter)
+                    )
+
+        def _validate_valid_at():
+            if self.valid_at:
+                if not (self.valid_at >= self.cert.get_notBefore() and self.valid_at <= self.cert.get_notAfter()):
+                    self.message.append(
+                        'Certificate is not valid for the specified date (%s) - notBefore: %s - notAfter: %s' % (self.valid_at, self.cert.get_notBefore(), self.cert.get_notAfter())
+                    )
+
+        def _validate_invalid_at():
+            if self.invalid_at:
+                if not (self.invalid_at <= self.cert.get_notBefore() or self.invalid_at >= self.cert.get_notAfter()):
+                    self.message.append(
+                        'Certificate is not invalid for the specified date (%s) - notBefore: %s - notAfter: %s' % (self.invalid_at, self.cert.get_notBefore(), self.cert.get_notAfter())
+                    )
+
+        def _validate_valid_in():
+            if self.valid_in:
+                valid_in_date = datetime.datetime.utcnow() + datetime.timedelta(seconds=self.valid_in)
+                valid_in_date = valid_in_date.strftime('%Y%m%d%H%M%SZ')
+                if not (valid_in_date >= self.cert.get_notBefore() and valid_in_date <= self.cert.get_notAfter()):
+                    self.message.append(
+                        'Certificate is not valid in %s seconds from now (%s) - notBefore: %s - notAfter: %s' % (self.valid_in, valid_in_date, self.cert.get_notBefore(), self.cert.get_notAfter())
+                    )
+
+        _validate_signature_algorithms()
+        _validate_subject()
+        _validate_issuer()
+        _validate_has_expired()
+        _validate_version()
+        _validate_keyUsage()
+        _validate_extendedKeyUsage()
+        _validate_subjectAltName()
+        _validate_notBefore()
+        _validate_notAfter()
+        _validate_valid_at()
+        _validate_invalid_at()
+        _validate_valid_in()
+
+        return len(self.message)
```

Comment:
A bit implicit for my taste, but the module is small and relatively self-contained. Still, I'd prefer returning at least a boolean.